File: Algorithms/127.WordLadder.py
class Solution(object):
    def ladderLength_bfs(self, beginWord, endWord, wordList):
        """
        :type beginWord: str
        :type endWord: str
        :type wordList: List[str]
        :rtype: int
        """
        wordDict = set(wordList)
        if endWord not in wordDict:
            return 0
        queue, step = [beginWord], 0
        while queue:
            step += 1
            for j in range(len(queue)):
                tmp = queue.pop(0)
                for i in range(len(tmp)):
                    for c in string.lowercase:
                        if c == tmp[i]: continue
                        temp = tmp[:i] + c + tmp[i+1:]
                        if temp == endWord:
                            return step + 1
                        if temp not in wordDict:
                            continue
                        wordDict.remove(temp)
                        queue.append(temp)
        return 0

    def ladderLength(self, beginWord, endWord, wordList):
        """
        :type beginWord: str
        :type endWord: str
        :type wordList: List[str]
        :rtype: int
        """
        wordDict = set(wordList)
        if endWord not in wordDict:
            return 0
        q1, q2, step = {beginWord}, {endWord}, 0
        wordDict.remove(endWord)
        while q1 and q2:
            step += 1
            if len(q1) > len(q2):
                q1, q2 = q2, q1
            q = set()
            for j in range(len(q1)):
                tmp = q1.pop()
                for i in range(len(tmp)):
                    for c in string.lowercase:
                        if c == tmp[i]: continue
                        temp = tmp[:i] + c + tmp[i+1:]
                        if temp in q2:
                            return step + 1
                        if temp not in wordDict:
                            continue
                        wordDict.remove(temp)
                        q.add(temp)
            q1 = q
        return 0

File: Algorithms/cpp/127.WordLadder.cpp
// https://zxi.mytechroad.com/blog/searching/127-word-ladder/
class Solution {
public:
    /*
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        unordered_set<string> dict(wordList.begin(), wordList.end());        
        if (!dict.count(endWord)) return 0;
        queue<string> q;
        q.push(beginWord);
        unordered_set<string> visited;
        visited.insert(beginWord);
        int l = beginWord.length();
        int step = 0;
        while (!q.empty()) {
            step ++;
            for (int size=q.size(); size > 0; size--) {
                string tmp = q.front();
                q.pop();
                for (int i=0; i < l; i++) {
                    char ch = tmp[i];
                    for (char c='a'; c <= 'z'; c++) {
                        if (c == tmp[i]) 
                            continue;
                        tmp[i] = c;
                        if (tmp == endWord) {
                            return step + 1;
                        }
                        if (!dict.count(tmp)) 
                            continue;
                        dict.erase(tmp);
                        q.push(tmp);
                    }
                    tmp[i] = ch;
                }
            }
        }
        return 0;
    }
    */
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        unordered_set<string> dict(wordList.begin(), wordList.end());
        if (!dict.count(endWord)) return 0;

        int l = beginWord.length();

        unordered_set<string> q1{beginWord};
        unordered_set<string> q2{endWord};

        int step = 0;

        while (!q1.empty() && !q2.empty()) {
            ++step;

            // Always expend the smaller queue first
            if (q1.size() > q2.size())
                std::swap(q1, q2);

            unordered_set<string> q;

            for (string w : q1) {
                for (int i = 0; i < l; i++) {
                    char ch = w[i];
                    for (int j = 'a'; j <= 'z'; j++) {
                        if (j == ch) 
                            continue;
                        w[i] = j;
                        if (q2.count(w)) return step + 1;
                        if (!dict.count(w)) continue;
                        dict.erase(w);
                        q.insert(w);
                    }
                    w[i] = ch;
                }
            }

            std::swap(q, q1);
        }
        return 0;
    }
};
